Add a Player

gustafla opened the pull request:
I hacked together a player and save_tracks -functionality with serde and bincode. Seems to work on my machine :tm:. Improvement suggestions are welcome, and you can refuse to merge if you see any issues. I know the player's use of HashMap is a bit questionable, I really should profile it but I have a demo to make :man_shrugging:. Kind of unsuitable for 4ks and the like because of dependencies on std and error handling bloat :D
Thanks and have a nice week

gustafla committed:
Add serde and bincode, Client::save_tracks

gustafla committed:
Add SaveTracksError and improve documentation

gustafla committed:
Add tracks.bin (from example.rs) to .gitignore

gustafla committed:
Make save_tracks public and no longer auto-save

gustafla committed:
Improve example code

gustafla committed:
Add a simple Player and a common trait (Rocket)

gustafla committed:
Remove trait Rocket

gustafla committed:
Improve documentation

gustafla committed:
Fix readme

gustafla committed:
Improve documentation

Darksecond commented on src/client.rs:
```diff
@@ -48,55 +62,55 @@ enum ReceiveResult {
-/// The `Rocket` type. This contains the connected socket and other fields.
-pub struct Rocket {
+/// The `Client` type. This contains the connected socket and other fields.
+pub struct Client {
```
Maybe call this `RocketClient` instead.

Darksecond commented on src/player.rs:
```diff
@@ -0,0 +1,47 @@
+///
+/// After constructing, call [`Player::get_track`] to get tracks.
+/// Then call [`Track::get_value`] to get saved values at any given point in time.
+pub struct Player {
```
This could then be called `RocketPlayer`.

gustafla replied:
Yeah when naming types like crate::CrateType, after `use`ing, the type still tells everything about itself. Even though it's a bit redundant. I think your suggestion is the best option in this case

Darksecond commented on src/player.rs:
```diff
@@ -0,0 +1,47 @@
+        // Load from file
+        let file = File::open(path).map_err(Error::OpenTrackFile)?;
+        let tracks_vec: Vec<Track> =
+            bincode::deserialize_from(file).map_err(Error::DeserializeTracks)?;
```
Do we want to decide for the user what format they want to serialize in, and if we do, do we want to document this somewhere?

gustafla replied:
Another solution I can think of woud be to remove bincode entirely and make `save_tracks` and `RocketPlayer::new` just return and take a `Vec<Track>` (`Track` still implementing `Serialize` and `Deserialize` of course). Then the demo could decide how to store the tracks. Or maybe offer both options, `RocketClient::get_tracks` and `RocketPlayer::from_tracks` but also feature-gated `RocketClient::save_tracks` (maybe call the feature bincode). Also, serde itself could be a toggleable feature but does anyone ever __not__ wanna be able to save and load? Then there is also the possibility of taking Rocket editor XML-ish files, but I don't know if other editors support that format

Darksecond replied:
I like the idea of implementing `Serialize` and `Deserialize` of course. I do prefer not having bincode as a dependency at this stage, I'd rather have a `to_tracks` or something to turn a `RocketClient` into a Serializable thing, and have a `from_tracks` (or new) on `RocketPlayer`.

gustafla replied:
If Tracks (or RocketData or something) was a type with just `Vec<Track>` inside, we could `impl From<RocketData> for Player` which I think is kinda neat (even though `from_tracks(Vec<Track>)` would work fine too). I'm just not sure if it's a good idea to turn `RocketClient` __into__ anything because wouldn't that then make it necessary to reconstruct the Client if the user wishes to keep editing? I think just giving the ability for the client to dump clones of it's Vec<Track> which can then be serialized regardless if the Rocket session is kept or destroyed would suit my use case better. I can also expand the examples to show reading and writing files (moving bincode to dev-dependencies for it to work)

Darksecond replied:
I think just cloning the Tracks is a more straightforward solution for now. As for bincode, it would be excellent to move that to dev deps and update the examples.

gustafla replied:
Hmm or, what if the Client can be straight up turned into a Player and the Player can be (de)serialized?

gustafla replied:
No, that would require the Client to have access to Player's internals or the Player to have a public contructor anyway. I think we'll roll with just save_tracks giving clones of Vec<Track> and Player::new taking Vec<Track>s. Thanks for the feedback again

Darksecond commented:
There are a few small things I'm not sure about and could be thought over. In general I do like the idea of adding playing abilities.

gustafla committed:
Rename main types to RocketClient and RocketPlayer

gustafla commented:
Thanks for the feedback. Yeah direct commit access would be appreciated, I plan to use this crate for a while because I'm not yet ready to develop my own demotool.

gustafla committed:
Remove dependency on bincode

This commit allows the user to choose their track serialization format.
Instead of providing file RW, the crate now only gives and consumes
clones of internal state required for playback in release builds.
